mindmeld/resource_loader.py

```diff
@@ -40,7 +40,105 @@
 
 logger = logging.getLogger(__name__)
 
-ENABLE_STEMMING_ARGS = "enable_stemming"
+
+class ProcessedQueryList:
+    """
+    ProcessedQueryList provides a memory efficient disk backed list representation
+    for a list of queries.
+    """
+    def __init__(self, cache=None, row_ids=None):
+        self._cache = cache
+        self.row_ids = row_ids or []
+
+    @property
+    def cache(self):
+        return self._cache
+
+    @cache.setter
+    def cache(self, cache):
+        self._cache = cache
+
+    def append(self, query_id):
+        self.row_ids.append(query_id)
+
+    def extend(self, query_ids):
+        self.row_ids.extend(query_ids)
+
+    def __len__(self):
+        return len(self.row_ids)
+
+    def __bool__(self):
+        return bool(len(self))
+
+    def processed_queries(self):
+        return ProcessedQueryList.Iterator(self)
+
+    def raw_queries(self):
+        return ProcessedQueryList.RawQueryIterator(self)
+
+    def queries(self):
+        return ProcessedQueryList.QueryIterator(self)
+
+    def entities(self):
+        return ProcessedQueryList.EntitiesIterator(self)
+
+    def domains(self):
+        return ProcessedQueryList.DomainIterator(self)
+
+    def intents(self):
+        return ProcessedQueryList.IntentIterator(self)
+
+    class Iterator:
+        def __init__(self, source):
+            self.source = source
+            self.row_ids = source.row_ids
+            self.iter_idx = -1
+
+        def __iter__(self):
+            self.iter_idx = -1
+            return self
+
+        def __next__(self):
+            self.iter_idx += 1
+            if self.iter_idx == len(self.source.row_ids):
+                raise StopIteration
+            return self[self.iter_idx]
+
+        def __len__(self):
+            return len(self.source)
+
+        def __getitem__(self, key):
+            '''
+            Override this function for getting other types of cached data.
+            '''
+            return self.source.cache.get(self.row_ids[key])
+
+        @staticmethod
+        def passthrough(processed_query):
+            return processed_query
+
+        def reorder(self, indices):
+            self.row_ids = [self.row_ids[i] for i in indices]
```

Comment:
Is the expectation that the indices passed in are sorted according to the values to match the alternate `.sorted()` operation? I am struggling to find where this is called in this UI, but from my memory, we passed in `.reoder([i in range(len(list))])` and the alternate was `list.sorted()`. I could be mistaken.

Reply:
This is called from the `fit()` method in tagger_models and text_models. They create a list of integers from 0-len(list), randomize it, then use it to randomize the examples and labels in the same order.